### PR TITLE
feat: Add variable substitution on ConfigMapKeySelector. Fixes #7061

### LINKED ISSUE #7061: Variable substitution does not work for `ConfigMapKeySelector` at global `Workflow` or `Step` level
### Environment

* Latest `argo-workflows` from `master` installed from https://raw.githubusercontent.com/argoproj/argo-workflows/master/manifests/quick-start-postgres.yaml
* Kubernetes version: `v1.22.2`
* Minikube version: `v1.23.2`
* Mac OS X 11.6

Given the instructions defined here https://github.com/argoproj/argo-workflows/blob/master/docs/variables.md I expect that variables work in every field of a `Workflow` resource. I am using a `ConfigMapKeySelector` which has a key that references a global or local step variable. The expansion does not work and the following template is parsed as a string:
`{{ workflow.parameters.var }}`

## Diagnostics

Example `ConfigMap` applied on the `argo` namespace:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: workflow-template-bug
  namespace: argo
  labels:
    # Note that this label is required for the informer to detect this ConfigMap.
    workflows.argoproj.io/configmap-type: Parameter
data:
  testKey: 'testValue'
```

Applied `Workflow`s:

```yaml
metadata:
  name: delightful-python
  namespace: argo
spec:
  arguments:
    parameters:
      - name: reference
        value: testKey
      - name: message
        valueFrom:
          configMapKeyRef:
            name: workflow-template-bug
            key: '{{ workflow.parameters.reference }}'
  entrypoint: argosay
  templates:
    - name: argosay
      inputs:
        parameters:
          - name: message
            value: '{{workflow.parameters.message}}'
      container:
        name: main
        image: 'argoproj/argosay:v2'
        command:
          - /argosay
        args:
          - echo
          - '{{inputs.parameters.message}}'

# or 

metadata:
  name: delightful-python
  namespace: argo
  labels:
    example: 'true'
spec:
  arguments:
    parameters:
      - name: reference
        value: testKey
  entrypoint: argosay
  templates:
    - name: argosay
      inputs:
        parameters:
          - name: message
            valueFrom:
              configMapKeyRef:
                name: workflow-template-bug
                key: '{{ workflow.parameters.reference }}'
      container:
        name: main
        image: 'argoproj/argosay:v2'
        command:
          - /argosay
        args:
          - echo
          - '{{inputs.parameters.message}}'

```

The error messages I get when submitting this `Workflow` from the UI are:

> Failed: failed to set global parameters: failed to set global parameter message from configmap with name workflow-template-bug and key {{workflow.parameters.reference}}: ConfigMap 'workflow-template-bug' does not have the key '{{workflow.parameters.reference}}'

and for the second `Workflow`:

> unable to retrieve inputs.parameters.message from ConfigMap: ConfigMap 'workflow-template-bug' does not have the key '{{ workflow.parameters.reference }}'

Is this a problem with nested fields or doesn't the `ConfigMapKeySelector` work with dynamic variable substitution?

Linking to argoproj/argo-workflows#6869

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by yriveiro**
Signed-off-by: Yago Riveiro <yago.riveiro@gmail.com>

This PR adds support for variable substitution on `ConfigMapKeyRef` allowing the following pattern:

```yaml
metadata:
  name: delightful-python
  namespace: argo
  labels:
    example: 'true'
spec:
  arguments:
    parameters:
      - name: reference
        value: testKey
  entrypoint: argosay
  templates:
    - name: argosay
      inputs:
        parameters:
          - name: message
            valueFrom:
              configMapKeyRef:
                name: workflow-template-bug
                key: '{{ workflow.parameters.reference }}'
```
or 

```yaml
metadata:
  name: delightful-python
  namespace: argo
  labels:
    example: 'true'
spec:
  arguments:
    parameters:
      - name: reference
        value: testName
  entrypoint: argosay
  templates:
    - name: argosay
      inputs:
        parameters:
          - name: message
            valueFrom:
              configMapKeyRef:
                name: '{{ workflow.parameters.reference }}'
                key: workflow-template-bug
```
I submitted the workflow using `argo submit` and see if the substitution were done.

This PR implements #7061 feature requests.

## Pre-requisites:
* [x] Run `make pre-commit -B`

### CODE REVIEW

**Review by terrytangyuan — CHANGES_REQUESTED** [review]
We probably need some test cases (e2e tests if unit tests are not trivial to implement)

**Comment by terrytangyuan** [comment]
@sarabala1979 Can I get your review on this as well?

**Comment by sarabala1979** [comment]
@yriveiro can you add a test for this?

**Comment by yriveiro** [comment]
> @yriveiro can you add a test for this?

Could someone help me with that? 

It's my first PR in the project and I don't know all the wiring to test this properly.

**Comment by sarabala1979** [comment]
@terrytangyuan can you help him write the test?

**Comment by terrytangyuan** [comment]
@yriveiro You can just add one additional example and then mark it as part of the existing e2e test via `workflows.argoproj.io/test: "true"` and check the status `workflows.argoproj.io/verify.py` similar to this: https://github.com/terrytangyuan/argo-workflows/blob/c2bcfb1fff231ae82e0f904691a76bda330c3236/examples/global-parameters-from-configmap.yaml

File: workflow/common/util.go [hunk 1]
```diff
@@ -183,6 +186,26 @@ func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams,
...
+		if v, ok := globalParams[k]; ok {
+			name = v
+		}
+	}
+	if strings.HasPrefix(key, "{{") && strings.HasPrefix(key, "}}") {
+		k := strings.TrimSuffix(strings.TrimPrefix(inParam.ValueFrom.ConfigMapKeyRef.Key, "{{"), "}}")
+		if v, ok := globalParams[k]; ok {
+			key = v
+		}
+	}
```
Comment (terrytangyuan) [inline workflow/common/util.go:205]: Redundant?

Comment (yriveiro) [reply workflow/common/util.go]: I refactored a bit the code to reduce redundancy.

**Comment by yriveiro** [comment]
> @yriveiro You can just add one additional example and then mark it as part of the existing e2e test via `workflows.argoproj.io/test: "true"` and check the status `workflows.argoproj.io/verify.py` similar to this: https://github.com/terrytangyuan/argo-workflows/blob/c2bcfb1fff231ae82e0f904691a76bda330c3236/examples/global-parameters-from-configmap.yaml

I added some tests, let me know if they are ok.

File: workflow/common/util.go [hunk 2]
```diff
@@ -183,6 +196,23 @@ func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams,
 	return SubstituteParams(newTmpl, globalParams, localParams)
 }
 
+// substituteConfigMapKeyRefParams check if ConfigMapKeyRef's key is a param and perform the substitution.
```
Comment (sarabala1979) [inline workflow/common/util.go:199]: can you add unit test for this?

**Comment by sarabala1979** [comment]
@yriveiro Can you add test and fix the unit test?

**Comment by yriveiro** [comment]
@sarabala1979 one of the E2E tests seems to fail due to k3s installation. I didn't find any place to retrigger only that test.

**Comment by terrytangyuan** [comment]
> @sarabala1979 one of the E2E tests seems to fail due to k3s installation. I didn't find any place to retrigger only that test.

I just re-triggered the tests.

**Comment by yriveiro** [comment]
> > @sarabala1979 one of the E2E tests seems to fail due to k3s installation. I didn't find any place to retrigger only that test.
> 
> I just re-triggered the tests.

No luck, the test get cancelled after 20 minutes with:

```
No resources found
Running examples/arguments-parameters-from-configmap.yaml...
workflow.argoproj.io/arguments-parameters-from-configmap-kvzz2 created
error: timed out waiting for the condition on workflows/arguments-parameters-from-configmap-kvzz2
make: *** [Makefile:504: test-examples] Error 1
```

**Comment by terrytangyuan** [comment]
Seems like a different test (examples) has failed this time. I just re-triggered again. Perhaps you could try run the example locally to see if it's stuck in the meantime?

**Pull request merged**